ospf6d/ospf6_asbr.c

```diff
@@ -2613,3 +2693,990 @@ void install_element_ospf6_debug_asbr(void)
+	}
+}
+
+static int ospf6_asbr_summary_process(struct thread *thread)
+{
+	struct ospf6 *ospf6 = THREAD_ARG(thread);
+	int operation = 0;
+
+	ospf6->t_external_aggr = NULL;
+	operation = ospf6->aggr_action;
+
+	if (IS_OSPF6_DEBUG_AGGR)
+		zlog_debug("%s: operation:%d",
+				__func__,
+				operation);
+
+	switch (operation) {
+	case OSPF6_ROUTE_AGGR_ADD:
+		ospf6_handle_external_aggr_add(ospf6);
+		break;
+	case OSPF6_ROUTE_AGGR_DEL:
+	case OSPF6_ROUTE_AGGR_MODIFY:
+		ospf6_handle_external_aggr_update(ospf6);
+		break;
+	default:
+		break;
+	}
+
+	return OSPF6_SUCCESS;
+}
+
+static void
+ospf6_start_asbr_summary_delay_timer(struct ospf6 *ospf6,
+			struct ospf6_external_aggr_rt *aggr,
+			ospf6_aggr_action_t operation)
+{
+	aggr->action = operation;
+
+	if (ospf6->t_external_aggr) {
+		if (ospf6->aggr_action == OSPF6_ROUTE_AGGR_ADD) {
+
+			if (IS_OSPF6_DEBUG_AGGR)
+				zlog_debug("%s: Not required to restart timer,set is already added.",
+					__func__);
+			return;
+		}
+
+		if (operation == OSPF6_ROUTE_AGGR_ADD) {
+			if (IS_OSPF6_DEBUG_AGGR)
+				zlog_debug("%s, Restarting Aggregator delay timer.",
+							__func__);
+			THREAD_OFF(ospf6->t_external_aggr);
+		}
+	}
+
+	if (IS_OSPF6_DEBUG_AGGR)
+		zlog_debug("%s: Start Aggregator delay timer %d(in seconds).",
+			__func__, ospf6->aggr_delay_interval);
+
+	ospf6->aggr_action = operation;
+	thread_add_timer(master,
+			ospf6_asbr_summary_process,
+			ospf6, ospf6->aggr_delay_interval,
+			&ospf6->t_external_aggr);
+}
+
+int ospf6_asbr_external_rt_advertise(struct ospf6 *ospf6,
+				struct prefix *p)
+{
+	struct route_node *rn;
+	struct ospf6_external_aggr_rt *aggr;
+
+	rn = route_node_lookup(ospf6->rt_aggr_tbl, (struct prefix *)p);
+	if (!rn)
+		return OSPF6_INVALID;
+
+	aggr = rn->info;
+
+	route_unlock_node(rn);
+
+	if (!CHECK_FLAG(aggr->aggrflags, OSPF6_EXTERNAL_AGGRT_NO_ADVERTISE))
+		return OSPF6_INVALID;
+
+	UNSET_FLAG(aggr->aggrflags, OSPF6_EXTERNAL_AGGRT_NO_ADVERTISE);
+
+	if (!OSPF6_EXTERNAL_RT_COUNT(aggr))
+		return OSPF6_SUCCESS;
+
+	ospf6_start_asbr_summary_delay_timer(ospf6, aggr,
+					     OSPF6_ROUTE_AGGR_MODIFY);
+
+	return OSPF6_SUCCESS;
+}
+
+int ospf6_external_aggr_delay_timer_set(struct ospf6 *ospf6,
+			unsigned int interval)
+{
+	ospf6->aggr_delay_interval = interval;
+
+	return OSPF6_SUCCESS;
+}
+
+static unsigned int ospf6_external_rt_hash_key(const void *data)
+{
+	const struct ospf6_route *rt = data;
+	unsigned int key = 0;
+
+	key = prefix_hash_key(&rt->prefix);
+	return key;
+}
+
+static bool ospf6_external_rt_hash_cmp(const void *d1, const void *d2)
+{
+	const struct ospf6_route *rt1 = d1;
+	const struct ospf6_route *rt2 = d2;
+
+	return prefix_same(&rt1->prefix, &rt2->prefix);
+}
+
+static struct ospf6_external_aggr_rt *
+ospf6_external_aggr_new(struct prefix *p)
+{
+	struct ospf6_external_aggr_rt *aggr;
+
+	aggr = XCALLOC(MTYPE_OSPF6_EXTERNAL_RT_AGGR,
+		       sizeof(struct ospf6_external_aggr_rt));
+
+	aggr->p.family = p->family;
+	prefix_copy(&aggr->p, p);
```

Comment:
You don't need to manually copy `family` and `prefixlen` if using `prefix_copy`, it already does that:

```c
void prefix_copy(union prefixptr udest, union prefixconstptr usrc)
{
        struct prefix *dest = udest.p;
        const struct prefix *src = usrc.p;

        dest->family = src->family;
        dest->prefixlen = src->prefixlen;

        if (src->family == AF_INET)
                dest->u.prefix4 = src->u.prefix4;
        else if (src->family == AF_INET6)
                dest->u.prefix6 = src->u.prefix6;
        // ...
}
```